### PR TITLE
Add local search history to the community list screen

### PR DESCRIPTION
**Pull request by wkoomson**
This change adds a search history Dao/Repository/ViewModel so that we can save user searches in Room, and display them on the community list screen. Users can click the history item to search it, or click the "X" button to delete it.

This is for issue  #811

### CODE REVIEW

**Comment by wkoomson** [comment]
Uploaded screen cap of new functionality
https://github.com/dessalines/jerboa/assets/861971/babf157a-ce2b-4402-a9bb-ff483cef7cb9

**Comment by MV-GH** [comment]
Can you add a option to disable saving searches, I can imagine that not everyone wants their searches to be saved.

**Comment by twizmwazin** [comment]
Based on the screen capture, I think searches should be listed in reverse-chronological order. The last thing I searched should be the first item in history.

**Comment by wkoomson** [comment]
Sorted searches in reverse order, added setting to control whether searches are saved or not.

screen cap

https://github.com/dessalines/jerboa/assets/861971/85f17c7b-c1ec-4de3-bfbe-1440560e9693

**Comment by MV-GH** [comment]
If you disable the saved searches does it delete the ones it had saved already?

**Comment by wkoomson** [comment]
Updated to clear search history when the option is disabled.

File: app/src/main/java/com/jerboa/ui/components/community/list/CommunityListActivity.kt [hunk 1]
```diff
@@ -53,24 +56,96 @@ fun CommunityListActivity(
...
+                                .imePadding(),
+                        ) {
+                            val history by communityListViewModel.searchHistory.collectAsState(
+                                emptyList(),
+                            )
+                            if (history.isNotEmpty()) {
+                                ListItem(
+                                    headlineContent = {
+                                        Text(
+                                            text = "Recent searches",
```
Comment (twizmwazin) [inline app/src/main/java/com/jerboa/ui/components/community/list/CommunityListActivity.kt:84]: Strings in this file need to be localized, then LGTM

**Review by twizmwazin — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/jerboa/ui/components/community/list/CommunityListActivity.kt (hunk 1 above)
Comment (wkoomson) [reply app/src/main/java/com/jerboa/ui/components/community/list/CommunityListActivity.kt]: Added string resources.

**Comment by wkoomson** [comment]
Apologies, I accidentally closed this pull request. I will create a new one. :facepalm:

**Review by MV-GH — CHANGES_REQUESTED** [review]
I cant actually run this, causes crash at startup see stacktrace

<details><summary>stack trace</summary>
E  (1) near "timestamp": syntax error in "CREATE TABLE IF NOT EXISTS SearchHistory(
                                                                                                                        history TEXT PRIMARY KEY NOT NULL
                                                                                                                        timestamp INTEGER NOT NULL 
                                                                                                                    )"
2023-06-28 06:01:10.690  6138-6138  SQLiteLog               com.jerboa.debug                     E  (1) near "timestamp": syntax error in "CREATE TABLE IF NOT EXISTS SearchHistory(
                                                                                                                        history TEXT PRIMARY KEY NOT NULL
                                                                                                                        timestamp INTEGER NOT NULL 
                                                                                                                    )"
FATAL EXCEPTION: main
Process: com.jerboa.debug, PID: 6138
java.lang.RuntimeException: Unable to start activity ComponentInfo{com.jerboa.debug/com.jerboa.MainActivity}: android.database.sqlite.SQLiteException: near "timestamp": syntax error (code 1 SQLITE_ERROR): , while compiling: CREATE TABLE IF NOT EXISTS SearchHistory(
                    history TEXT PRIMARY KEY NOT NULL
                    timestamp INTEGER NOT NULL 
                )
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3635)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3792)
	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103)
	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2210)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7839)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
Caused by: android.database.sqlite.SQLiteException: near "timestamp": syntax error (code 1 SQLITE_ERROR): , while compiling: CREATE TABLE IF NOT EXISTS SearchHistory(
                    history TEXT PRIMARY KEY NOT NULL
                    timestamp INTEGER NOT NULL 
                )
	at android.database.sqlite.SQLiteConnection.nativePrepareStatement(Native Method)
	at android.database.sqlite.SQLiteConnection.acquirePreparedStatement(SQLiteConnection.java:1047)
	at android.database.sqlite.SQLiteConnection.prepare(SQLiteConnection.java:654)
	at android.database.sqlite.SQLiteSession.prepare(SQLiteSession.java:590)
	at android.database.sqlite.SQLiteProgram.<init>(SQLiteProgram.java:62)
	at android.database.sqlite.SQLiteStatement.<init>(SQLiteStatement.java:34)
	at android.database.sqlite.SQLiteDatabase.executeSql(SQLiteDatabase.java:1920)
	at android.database.sqlite.SQLiteDatabase.execSQL(SQLiteDatabase.java:1842)
	at androidx.sqlite.db.framework.FrameworkSQLiteDatabase.execSQL(FrameworkSQLiteDatabase.kt:246)
	at com.jerboa.db.AppDBKt$MIGRATION_17_18$1.migrate(AppDB.kt:527)
	at androidx.room.RoomOpenHelper.onUpgrade(RoomOpenHelper.kt:91)
	at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.onUpgrade(FrameworkSQLiteOpenHelper.kt:253)
	at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:416)
	at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:316)
	at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.getWritableOrReadableDatabase(FrameworkSQLiteOpenHelper.kt:232)
	at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.innerGetDatabase(FrameworkSQLiteOpenHelper.kt:190)
	at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.getSupportDatabase(FrameworkSQLiteOpenHelper.kt:151)
	at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper.getWritableDatabase(FrameworkSQLiteOpenHelper.kt:104)
	at androidx.room.RoomDatabase.inTransaction(RoomDatabase.kt:638)
	at androidx.room.RoomDatabase.assertNotSuspendingTransaction(RoomDatabase.kt:457)
	at com.jerboa.db.AccountDao_Impl.getAllSync(AccountDao_Impl.java:289)
	at com.jerboa.db.AccountRepository.getAllSync(AppDB.kt:227)
	at com.jerboa.db.AccountViewModel.<init>(AppDB.kt:610)
	at com.jerboa.db.AccountViewModelFactory.create(AppDB.kt:634)
	at androidx.lifecycle.ViewModelProvider$Factory.create(ViewModelProvider.kt:83)
	at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.kt:187)
	at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.kt:153)
	at androidx.lifecycle.ViewModelLazy.getValue(ViewModelLazy.kt:53)
	at androidx.lifecycle.ViewModelLazy.getValue(ViewModelLazy.kt:35)
	at com.jerboa.MainActivity.getAccountViewModel(MainActivity.kt:117)
	at com.jerboa.MainActivity.onCreate(MainActivity.kt:128)
	at android.app.Activity.performCreate(Activity.java:8051)
	at android.app.Activity.performCreate(Activity.java:8031)
	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1329)
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3608)
	... 12 more

hi

</details>

File: app/src/main/java/com/jerboa/MainActivity.kt [hunk 2]
```diff
@@ -79,11 +90,27 @@ import com.jerboa.ui.theme.JerboaTheme
...
+    private val personProfileViewModel by viewModels<PersonProfileViewModel>()
+    private val inboxViewModel by viewModels<InboxViewModel>()
+    private val communityListViewModel by viewModels<CommunityListViewModel>() {
+        CommunityListViewModelFactory((application as JerboaApplication).searchHistoryRepository)
+    }
+    private val createPostViewModel by viewModels<CreatePostViewModel>()
+    private val commentReplyViewModel by viewModels<CommentReplyViewModel>()
+    private val commentEditViewModel by viewModels<CommentEditViewModel>()
+    private val postEditViewModel by viewModels<PostEditViewModel>()
+    private val createReportViewModel by viewModels<CreateReportViewModel>()
```
Comment (MV-GH) [inline app/src/main/java/com/jerboa/MainActivity.kt:113]: Why are these added?

Comment (wkoomson) [reply app/src/main/java/com/jerboa/MainActivity.kt]: Wonky rebase, thanks for catching. Removed.

**Comment by wkoomson** [comment]
> I cant actually run this, causes crash at startup see stacktrace
> stack trace
> E (1) near "timestamp": syntax error in "CREATE TABLE IF NOT EXISTS SearchHistory( history TEXT PRIMARY KEY NOT NULL timestamp INTEGER NOT NULL )" 2023-06-28 06:01:10.690 6138-6138 SQLiteLog com.jerboa.debug E (1) near "timestamp": syntax error in "CREATE TABLE IF NOT EXISTS SearchHistory( history TEXT PRIMARY KEY NOT NULL timestamp INTEGER NOT NULL )" FATAL EXCEPTION: main Process: com.jerboa.debug, PID: 6138 java.lang.RuntimeException: Unable to start activity ComponentInfo{com.jerboa.debug/com.jerboa.MainActivity}: android.database.sqlite.SQLiteException: near "timestamp": syntax error (code 1 SQLITE_ERROR): , while compiling: CREATE TABLE IF NOT EXISTS SearchHistory( history TEXT PRIMARY KEY NOT NULL timestamp INTEGER NOT NULL ) at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3635) at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3792) at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103) at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135) at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95) at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2210) at android.os.Handler.dispatchMessage(Handler.java:106) at android.os.Looper.loopOnce(Looper.java:201) at android.os.Looper.loop(Looper.java:288) at android.app.ActivityThread.main(ActivityThread.java:7839) at java.lang.reflect.Method.invoke(Native Method) at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548) at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003) Caused by: android.database.sqlite.SQLiteException: near "timestamp": syntax error (code 1 SQLITE_ERROR): , while compiling: CREATE TABLE IF NOT EXISTS SearchHistory( history TEXT PRIMARY KEY NOT NULL timestamp INTEGER NOT NULL ) at android.database.sqlite.SQLiteConnection.nativePrepareStatement(Native Method) at android.database.sqlite.SQLiteConnection.acquirePreparedStatement(SQLiteConnection.java:1047) at android.database.sqlite.SQLiteConnection.prepare(SQLiteConnection.java:654) at android.database.sqlite.SQLiteSession.prepare(SQLiteSession.java:590) at android.database.sqlite.SQLiteProgram.(SQLiteProgram.java:62) at android.database.sqlite.SQLiteStatement.(SQLiteStatement.java:34) at android.database.sqlite.SQLiteDatabase.executeSql(SQLiteDatabase.java:1920) at android.database.sqlite.SQLiteDatabase.execSQL(SQLiteDatabase.java:1842) at androidx.sqlite.db.framework.FrameworkSQLiteDatabase.execSQL(FrameworkSQLiteDatabase.kt:246) at com.jerboa.db.AppDBKt$MIGRATION_17_18$1.migrate(AppDB.kt:527) at androidx.room.RoomOpenHelper.onUpgrade(RoomOpenHelper.kt:91) at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.onUpgrade(FrameworkSQLiteOpenHelper.kt:253) at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:416) at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:316) at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.getWritableOrReadableDatabase(FrameworkSQLiteOpenHelper.kt:232) at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.innerGetDatabase(FrameworkSQLiteOpenHelper.kt:190) at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.getSupportDatabase(FrameworkSQLiteOpenHelper.kt:151) at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper.getWritableDatabase(FrameworkSQLiteOpenHelper.kt:104) at androidx.room.RoomDatabase.inTransaction(RoomDatabase.kt:638) at androidx.room.RoomDatabase.assertNotSuspendingTransaction(RoomDatabase.kt:457) at com.jerboa.db.AccountDao_Impl.getAllSync(AccountDao_Impl.java:289) at com.jerboa.db.AccountRepository.getAllSync(AppDB.kt:227) at com.jerboa.db.AccountViewModel.(AppDB.kt:610) at com.jerboa.db.AccountViewModelFactory.create(AppDB.kt:634) at androidx.lifecycle.ViewModelProvider$Factory.create(ViewModelProvider.kt:83) at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.kt:187) at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.kt:153) at androidx.lifecycle.ViewModelLazy.getValue(ViewModelLazy.kt:53) at androidx.lifecycle.ViewModelLazy.getValue(ViewModelLazy.kt:35) at com.jerboa.MainActivity.getAccountViewModel(MainActivity.kt:117) at com.jerboa.MainActivity.onCreate(MainActivity.kt:128) at android.app.Activity.performCreate(Activity.java:8051) at android.app.Activity.performCreate(Activity.java:8031) at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1329) at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3608) ... 12 more
> 
> ```sql
>  CREATE TABLE IF NOT EXISTS SearchHistory(
>                     history TEXT PRIMARY KEY NOT NULL
>                     timestamp INTEGER NOT NULL 
>                 )
> ```
> 
> its missing a comma between history and timestamp

This should be working now, thanks for catching this.

**Comment by MV-GH** [comment]
It still crashes at startup but for a different reason this time. The best way to test the migration is to install a version with a DB behind the current. and then install the new app.


<details><summary>The stacktrace now is</summary>

  Process: com.jerboa.debug, PID: 12543
                                                                                                    java.lang.RuntimeException: Unable to start activity ComponentInfo{com.jerboa.debug/com.jerboa.MainActivity}: java.lang.IllegalStateException: Migration didn't properly handle: SearchHistory(com.jerboa.db.SearchHistory).
                                                                                                     Expected:
                                                                                                    TableInfo{name='SearchHistory', columns={text=Column{name='text', type='TEXT', affinity='2', notNull=true, primaryKeyPosition=1, defaultValue='undefined'}, timestamp=Column{name='timestamp', type='INTEGER', affinity='3', notNull=true, primaryKeyPosition=0, defaultValue='undefined'}}, foreignKeys=[], indices=[]}
                                                                                                     Found:
                                                                                                    TableInfo{name='SearchHistory', columns={history=Column{name='history', type='TEXT', affinity='2', notNull=true, primaryKeyPosition=1, defaultValue='undefined'}, timestamp=Column{name='timestamp', type='INTEGER', affinity='3', notNull=true, primaryKeyPosition=0, defaultValue='undefined'}}, foreignKeys=[], indices=[]}
</details>

And this is because in your migration your column is named history.  But your entity the column has name "text".

```kt
|@Entity
data class SearchHistory(
    @PrimaryKey @ColumnInfo(name = "text") val text: String, // Should be history or migration should be text
    @ColumnInfo(name = "timestamp") val timestamp: Long,
)
```

File: app/src/main/java/com/jerboa/db/AppDB.kt [hunk 3]
```diff
@@ -142,8 +149,30 @@ val APP_SETTINGS_DEFAULT = AppSettings(
     usePrivateTabs = false,
     secureWindow = false,
     blurNSFW = true,
+    saveSearchHistory = true,
+)
+
+@Entity
+data class SearchHistory(
+    @PrimaryKey @ColumnInfo(name = "text") val text: String,
```
Comment (dessalines) [inline app/src/main/java/com/jerboa/db/AppDB.kt:157]: No, use auto-gen'd primary keys like the other example.

2ndly, don't name these columns names which might be protected keywords in sql or kotlin. Name it `search_term` or something.

3rd: These search terms should almost certainly be attached to an account, so have a foreign key link to account.id

File: app/src/main/java/com/jerboa/db/AppDB.kt [hunk 4]
```diff
@@ -142,8 +149,30 @@ val APP_SETTINGS_DEFAULT = AppSettings(
...
+    @ColumnInfo(name = "timestamp") val timestamp: Long,
```
Comment (dessalines) [inline app/src/main/java/com/jerboa/db/AppDB.kt:158]: This column isn't necessary, the inserts are going to be in order anyway by the primary key.

File: app/src/main/java/com/jerboa/db/AppDB.kt [hunk 5]
```diff
@@ -173,6 +202,9 @@ interface AppSettingsDao {
     @Query("SELECT * FROM AppSettings limit 1")
     fun getSettings(): LiveData<AppSettings>
 
+    @Query("SELECT * FROM AppSettings limit 1")
```
Comment (dessalines) [inline app/src/main/java/com/jerboa/db/AppDB.kt:205]: Why did this get added? The livedata is already above this.

File: app/src/main/java/com/jerboa/db/AppDB.kt [hunk 6]
```diff
@@ -142,8 +149,30 @@ val APP_SETTINGS_DEFAULT = AppSettings(
...
+@Entity
+data class SearchHistory(
+    @PrimaryKey @ColumnInfo(name = "text") val text: String,
+    @ColumnInfo(name = "timestamp") val timestamp: Long,
 )
 
+@Dao
+interface SearchHistoryDao {
+    @Query("SELECT * FROM SearchHistory")
+    fun history(): Flow<List<SearchHistory>>
```
Comment (dessalines) [inline app/src/main/java/com/jerboa/db/AppDB.kt:164]: Use `LiveData` like the other examples. Don't code from scratch: use the conventions currently here.

File: app/src/main/java/com/jerboa/db/AppDB.kt [hunk 7]
```diff
@@ -224,10 +256,26 @@ class AccountRepository(private val accountDao: AccountDao) {
...
 }
 
+class SearchHistoryRepository(
+    private val searchHistoryDao: SearchHistoryDao,
+    private val appSettingsDao: AppSettingsDao,
+) {
+    fun history(): Flow<List<SearchHistory>> = searchHistoryDao.history()
+
+    suspend fun insert(item: SearchHistory) {
+        if (appSettingsDao.settings().first().saveSearchHistory) {
```
Comment (dessalines) [inline app/src/main/java/com/jerboa/db/AppDB.kt:266]: Why is this necessary? It should be connected to an account, not the appsettings, so it'll already have the account.id in the form.

There's also no need to do a check on this insert. If you have an appsetting to disable search, that should be done outside of / before this SQL logic.

File: app/src/main/java/com/jerboa/db/AppDB.kt [hunk 8]
```diff
@@ -242,6 +290,7 @@ class AppSettingsRepository(
     @WorkerThread
     suspend fun update(appSettings: AppSettings) {
         appSettingsDao.updateAppSettings(appSettings)
+        if (!appSettings.saveSearchHistory) searchHistoryDao.clear()
```
Comment (dessalines) [inline app/src/main/java/com/jerboa/db/AppDB.kt:293]: IMO this isn't necessary, but its up to you. The saved search history doesn't really need to clear anything, it just needs to visually not show it.

File: app/src/main/java/com/jerboa/db/AppDB.kt [hunk 9]
```diff
@@ -469,14 +518,32 @@ val MIGRATION_16_17 = object : Migration(16, 17) {
...
+val MIGRATION_17_18 = object : Migration(17, 18) {
+    override fun migrate(database: SupportSQLiteDatabase) {
+        database.execSQL(UPDATE_APP_CHANGELOG_UNVIEWED)
+        database.execSQL(
+            "ALTER TABLE AppSettings add column save_search_history INTEGER NOT NULL default 1",
+        )
+        database.execSQL(
+            """
+                CREATE TABLE IF NOT EXISTS SearchHistory(
+                    history TEXT PRIMARY KEY NOT NULL,
```
Comment (dessalines) [inline app/src/main/java/com/jerboa/db/AppDB.kt:530]: See above. Primary key issue, column names, and unecessary timestamp column.

File: app/src/main/java/com/jerboa/MainActivity.kt [hunk 10]
```diff
@@ -79,11 +82,19 @@ import com.jerboa.ui.theme.JerboaTheme
...
+    }
+    val searchHistoryRepository by lazy {
+        SearchHistoryRepository(database.searchHistoryDao(), database.appSettingsDao())
+    }
 }
 
 class MainActivity : AppCompatActivity() {
     private val siteViewModel by viewModels<SiteViewModel>()
+    private val communityListViewModel by viewModels<CommunityListViewModel>() {
+        CommunityListViewModelFactory((application as JerboaApplication).searchHistoryRepository)
```
Comment (dessalines) [inline app/src/main/java/com/jerboa/MainActivity.kt:96]: Why did you not make the searchhistory a viewmodel like the other tables? There's no reason to try to bundle it into another one.

File: app/src/main/java/com/jerboa/ui/components/community/list/CommunityListViewModel.kt [hunk 11]
```diff
@@ -4,34 +4,99 @@ import androidx.compose.runtime.getValue
...
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.map
 import kotlinx.coroutines.launch
+import java.time.Instant
 
-class CommunityListViewModel : ViewModel(), Initializable {
-    override var initialized by mutableStateOf(false)
+class CommunityListViewModelFactory(private val repository: SearchHistoryRepository) :
```
Comment (dessalines) [inline app/src/main/java/com/jerboa/ui/components/community/list/CommunityListViewModel.kt:33]: No, just make the searchhistory its own viewmodel like the other examples, and pass in either the viewmodel, or the `searchHistoryViewModel.observeAsState()` into the activities.

**Review by dessalines — CHANGES_REQUESTED** [review]
This needs a lot of work, because you didn't look at how the other tables were done, and tried to build it from scratch. The SearchHistory table needs some work, as well as its own viewmodel like the other two tables, that can be passed into activities as necessary.

File: app/src/main/java/com/jerboa/db/AppDB.kt (hunk 8 above)
Comment (MV-GH) [reply app/src/main/java/com/jerboa/db/AppDB.kt:293]: I wanted it, to be added because if you don't want your searches be saved it shouldn't still have them saved in the DB

**Comment by wkoomson** [comment]
Sounds good, I'll add another change shortly to make search history its own
view model and address the rest of the feedback.

One question, I added the timestamp column because there was a request to
display history entries sorted in reverse chronological order. Is this
still desired? If not I can remove that column.

File: app/src/main/java/com/jerboa/MainActivity.kt (hunk 10 above)
Comment (wkoomson) [reply app/src/main/java/com/jerboa/MainActivity.kt]: Added separate ViewModel

File: app/src/main/java/com/jerboa/db/AppDB.kt (hunk 8 above)
Comment (wkoomson) [reply app/src/main/java/com/jerboa/db/AppDB.kt]: I've kept this for now. Lmk if it needs to be removed,

File: app/src/main/res/values/strings.xml [hunk 12]
```diff
@@ -337,4 +337,8 @@
     <string name="back_dialog">Show a confirmation dialog on exit</string>
     <string name="exit">Exit</string>
     <string name="back_dialog_title">Are you sure you want to exit?</string>
+    <string name="community_list_recent_searches">Recent searches</string>
+    <string name="community_list_delete_search_item">Delete %1$s</string>
+    <string name="community_list_title">Communities\n</string>
+    <string name="save_search_history">Save search history\n</string>
```
Comment (MV-GH) [inline app/src/main/res/values/strings.xml:343]: Remove these newlines

**Review by MV-GH — CHANGES_REQUESTED** [review]
It is also missing the export 20 schema

File: app/src/main/java/com/jerboa/db/AppDBMigrations.kt [hunk 13]
```diff
@@ -261,6 +261,31 @@ val MIGRATION_19_20 = object : Migration(19, 20) {
...
+                    FOREIGN KEY(account_id)
+                        REFERENCES Account(id)
+                        ON DELETE CASCADE
+                        ON UPDATE CASCADE
+                )
+            """,
+        )
+        database.execSQL(
+            "CREATE INDEX index_SearchHistory_account_id ON SearchHistory(account_id)",
+        )
```
Comment (dessalines) [inline app/src/main/java/com/jerboa/db/AppDBMigrations.kt:285]: Looks much better, thanks.

File: app/src/main/java/com/jerboa/db/AppDBMigrations.kt [hunk 14]
```diff
@@ -261,6 +261,31 @@ val MIGRATION_19_20 = object : Migration(19, 20) {
...
+    override fun migrate(database: SupportSQLiteDatabase) {
+        database.execSQL(UPDATE_APP_CHANGELOG_UNVIEWED)
+        database.execSQL(
+            "ALTER TABLE AppSettings ADD COLUMN save_search_history INTEGER NOT NULL DEFAULT 1",
+        )
+        database.execSQL(
+            """
+                CREATE TABLE IF NOT EXISTS SearchHistory(
+                    id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, 
+                    account_id INTEGER,
```
Comment (dessalines) [inline app/src/main/java/com/jerboa/db/AppDBMigrations.kt:274]: account_id should also be not null, no? And it'd probably be clearer to move the references / fk relation into this line.

File: app/src/main/java/com/jerboa/db/dao/SearchHistoryDao.kt [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+@Dao
+interface SearchHistoryDao {
+    @Query("SELECT * FROM SearchHistory")
+    fun history(): LiveData<List<SearchHistory>>
+
+    @Insert(onConflict = OnConflictStrategy.REPLACE, entity = SearchHistory::class)
+    suspend fun insert(item: SearchHistory)
+
+    @Query("DELETE FROM SearchHistory WHERE account_id IS :accountId AND search_term = :searchTerm")
+    suspend fun delete(accountId: Int?, searchTerm: String)
```
Comment (dessalines) [inline app/src/main/java/com/jerboa/db/dao/SearchHistoryDao.kt:19]: account_id should be required.

File: app/src/main/java/com/jerboa/db/entity/SearchHistory.kt [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+    ],
+)
+data class SearchHistory(
+    @PrimaryKey(autoGenerate = true)
+    val id: Int = 0,
+    @ColumnInfo(
+        name = "account_id",
+        index = true,
+    )
+    val accountId: Int?,
```
Comment (dessalines) [inline app/src/main/java/com/jerboa/db/entity/SearchHistory.kt:26]: Same here, not optional.

File: app/src/main/java/com/jerboa/db/repository/SearchHistoryRepository.kt [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.map
+import com.jerboa.db.dao.SearchHistoryDao
+import com.jerboa.db.entity.SearchHistory
+
+class SearchHistoryRepository(
+    private val searchHistoryDao: SearchHistoryDao,
+) {
+    fun history(): LiveData<List<SearchHistory>> = searchHistoryDao.history()
+        .map { history ->
```
Comment (dessalines) [inline app/src/main/java/com/jerboa/db/repository/SearchHistoryRepository.kt:12]: Don't sort / distinct in code, sort using the `@Query` in sql.

File: app/src/main/java/com/jerboa/model/CommunityListViewModel.kt [hunk 18]
```diff
@@ -24,6 +23,8 @@ class CommunityListViewModel : ViewModel(), Initializable {
     var searchRes: ApiState<SearchResponse> by mutableStateOf(ApiState.Empty)
         private set
 
+    var communities: List<CommunityView> by mutableStateOf(emptyList())
```
Comment (dessalines) [inline app/src/main/java/com/jerboa/model/CommunityListViewModel.kt:26]: This should match main, looks like it was a wrong conflict fix.

File: app/src/main/java/com/jerboa/ui/components/community/list/CommunityListActivity.kt [hunk 19]
```diff
@@ -49,6 +57,8 @@ fun CommunityListActivity(
     siteViewModel: SiteViewModel,
     blurNSFW: Boolean,
     drawerState: DrawerState,
+    appSettingsViewModel: AppSettingsViewModel,
```
Comment (dessalines) [inline app/src/main/java/com/jerboa/ui/components/community/list/CommunityListActivity.kt:60]: Seems pointless to bring in the entire appSettingsViewModel here. Just add `saveSearchHistory` as a param.

File: app/src/main/java/com/jerboa/ui/components/community/list/CommunityListActivity.kt [hunk 20]
```diff
@@ -60,9 +70,20 @@ fun CommunityListActivity(
         communityListViewModel.setCommunityListFromFollowed(siteViewModel)
     }
 
+    val saveSearchHistory by remember {
```
Comment (dessalines) [inline app/src/main/java/com/jerboa/ui/components/community/list/CommunityListActivity.kt:73]: Not necessary.

File: app/src/main/java/com/jerboa/ui/components/community/list/CommunityListActivity.kt [hunk 21]
```diff
@@ -60,9 +70,20 @@ fun CommunityListActivity(
...
+        appSettingsViewModel.appSettings
+            .map { it.saveSearchHistory }
+    }.observeAsState()
+
+    val searchHistory by remember(account) {
+        searchHistoryViewModel.searchHistory
+            .map { history -> history.filter { it.accountId == account?.id } }
+    }.observeAsState()
```
Comment (dessalines) [inline app/src/main/java/com/jerboa/ui/components/community/list/CommunityListActivity.kt:81]: I thought there was a function that takes in `account_id` to filter by that already? If not, one needs to be added.

File: app/src/main/java/com/jerboa/db/AppDBContainer.kt [hunk 22]
```diff
@@ -3,9 +3,13 @@ package com.jerboa.db
...
+import com.jerboa.db.repository.SearchHistoryRepository
 
 class AppDBContainer(private val context: Context) {
     private val database by lazy { AppDB.getDatabase(context) }
     val accountRepository by lazy { AccountRepository(database.accountDao()) }
-    val appSettingsRepository by lazy { AppSettingsRepository(database.appSettingsDao()) }
+    val appSettingsRepository by lazy {
+        AppSettingsRepository(database.appSettingsDao(), database.searchHistoryDao())
+    }
+    val searchHistoryRepository by lazy { SearchHistoryRepository(database.searchHistoryDao()) }
```
Comment (dessalines) [inline app/src/main/java/com/jerboa/db/AppDBContainer.kt:14]: Just add this new one alongside the other two, there's no need to embed them.

File: app/src/main/java/com/jerboa/db/AppDBMigrations.kt [hunk 23]
```diff
@@ -261,6 +261,31 @@ val MIGRATION_19_20 = object : Migration(19, 20) {
...
+                    search_term TEXT NOT NULL,
+                    FOREIGN KEY(account_id)
+                        REFERENCES Account(id)
+                        ON DELETE CASCADE
+                        ON UPDATE CASCADE
+                )
+            """,
+        )
+        database.execSQL(
+            "CREATE INDEX index_SearchHistory_account_id ON SearchHistory(account_id)",
```
Comment (dessalines) [inline app/src/main/java/com/jerboa/db/AppDBMigrations.kt:284]: consistent naming: `idx_search_history_account`

**Comment by dessalines** [comment]
Stale PR. I can re-open if necessary.
